internal/vm/task.go

```diff
@@ -0,0 +1,335 @@
+	"github.com/containerd/containerd/log"
+	taskAPI "github.com/containerd/containerd/runtime/v2/task"
+	"github.com/containerd/fifo"
+	ptypes "github.com/gogo/protobuf/types"
+	"github.com/hashicorp/go-multierror"
+	"github.com/pkg/errors"
+
+	"github.com/firecracker-microvm/firecracker-containerd/internal"
+	"github.com/firecracker-microvm/firecracker-containerd/internal/bundle"
+	"github.com/firecracker-microvm/firecracker-containerd/proto"
+)
+
+// IODirection represents the direction in which container related I/O should
+// flow between a FIFO and a vsock connection (the possible directions being
+// from FIFO to vsock or from vsock to FIFO). This is needed to parameterize
+// functions that operate on either Host or Guest (in which case I/O directions
+// for the same stream are inverses) and functions that operate on stdin vs.
+// stdout/stderr (which again have inverse directions).
+type IODirection bool
+
+const (
+	// FIFOtoVSock represents I/O that should be copied from a FIFO to a VSock.
+	FIFOtoVSock IODirection = true
+
+	// VSockToFIFO represents I/O that should be copied from a VSock to a FIFO.
+	VSockToFIFO IODirection = false
+)
+
+func (d IODirection) opposite() IODirection {
+	return IODirection(!bool(d))
+}
+
+// VSockConnector is a function used as a callback for obtaining a vsock connection.
+// Implementations may include one that dials a remote vsock listener and one that
+// listens and accepts incoming connections.
+type VSockConnector func(ctx context.Context, port uint32) (net.Conn, error)
+
+// TaskManager manages a mapping of containerIDs to metadata for containerd tasks
+// being executed via a firecracker-containerd runtime. It's intended to be
+// abstracted over whether it's being executed on the Host or inside a VM Guest.
+type TaskManager struct {
+	mu    sync.RWMutex
+	tasks map[string]*Task
+}
+
+// AddTask registers a task with the provided metadata with the TaskManager.
+// taskService should implement the TaskService API for the task (i.e. Create, Kill, Exec, etc.).
+func (m *TaskManager) AddTask(
+	ctx context.Context,
+	containerID string,
+	taskService taskAPI.TaskService,
+	bundleDir bundle.Dir,
+	extraData *proto.ExtraData,
+	fifoSet *cio.FIFOSet,
+) (*Task, error) {
+	err := bundleDir.Create() // no-op if it already exists
+	if err != nil {
+		return nil, err
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if m.tasks == nil {
+		m.tasks = make(map[string]*Task)
+	}
+
+	if _, ok := m.tasks[containerID]; ok {
+		return nil, errors.Errorf("cannot add duplicate task with container ID %q", containerID)
+	}
+
+	task := &Task{
+		TaskService: taskService,
+		ID:          containerID,
+
+		extraData: extraData,
+		bundleDir: bundleDir,
+		fifoSet:   fifoSet,
+
+		cancelCh: make(chan struct{}),
+	}
+
+	m.tasks[task.ID] = task
+	return task, nil
+}
+
+func (m *TaskManager) get(containerID string) (*Task, error) {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	if m.tasks == nil {
+		m.tasks = make(map[string]*Task)
```

Comment:
nit: can you move this to `NewTaskManager` so we don't have to perform this checks in multiple places.